python/tank/util/login.py

```diff
@@ -65,35 +68,47 @@ def get_shotgun_user(sg):
         g_shotgun_user_cache = sg.find_one("HumanUser", filters=[["login", "is", local_login]], fields=fields)
     
     return g_shotgun_user_cache
-        
+
 def get_current_user(tk):
     """
-    Retrieves the current user as a dictionary of metadata values.
-    Returns None if the user is not found in shotgun.
-    
-    Returns the following fields:
-    
-    * id
-    * type
-    * email
-    * login
-    * name
-    * image url (thumbnail)
-    
-    This method connects to shotgun.
-    """    
+    Retrieves the current user as a dictionary of metadata values. Note: This method connects to
+    shotgun the first time around. The result is then cached to reduce latency.
+    :returns: None if the user is not found in shotgun. Otherwise, it returns a dictionary
+              with the following fields:
+                 * id
+                 * type
+                 * email
+                 * login
+                 * name
+                 * image url (thumbnail)
+    """
     global g_shotgun_current_user_cache
-    if g_shotgun_current_user_cache == "unknown":
-    
+    if g_shotgun_current_user_cache != "unknown":
+        return g_shotgun_current_user_cache
+
+    from .. import api
+
+    user = api.get_current_user()
+
+    if sg_auth.is_script_user(user):
+        # If we have a script user, try to find a matching user using the os user name.
         # call hook to get current login
         current_login = tk.execute_core_hook(constants.CURRENT_LOGIN_HOOK_NAME)
-        if current_login is None:
-            g_shotgun_current_user_cache = None
-        
-        else:
-            fields = ["id", "type", "email", "login", "name", "image"]
-            g_shotgun_current_user_cache = tk.shotgun.find_one("HumanUser", 
-                                                               filters=[["login", "is", current_login]], 
-                                                               fields=fields)
-    
+    elif sg_auth.is_session_user(user):
+        # If we have a human user, simply use the login value.
+        current_login = user.get_login()
+    else:
+        # Something is wrong, no current login available.
+        current_login = None
+
+    if current_login is None:
+        g_shotgun_current_user_cache = None
+    else:
+        fields = ["id", "type", "email", "login", "name", "image"]
+        g_shotgun_current_user_cache = tk.shotgun.find_one(
```

Comment:
Could this be cached on the user object as part of the authentication module? Seems odd to me that we have two globals in core storing two different kinds of current user! If we store it as part of the auth framework, we could also persist this metadata alongside the session ids etc.?